### DISCUSSION TITLE
_malloc/stackAlloc

### DISCUSSION THREAD

**Discussion by cemalgnlts** [emscripten-core/emscripten#21468] [Q&A] [answered]
Hi,
I was looking at the `stringToUTF8OnStack` function and it was sending the pointer from `stackAlloc` to the `stringToUTF8` function. Just like the `stringToNewUTF8` function does, but the `stringToNewUTF8` function does it with `_malloc`.

My question is that the `stringToUTF8` function redirects the incoming pointer to `HEAPU8`. So here the pointer value from `stackAlloc` and `_malloc` seems to be no different.

If this is the case, why do we have the responsibility to free the `_malloc` pointer but not the pointer from `stackAlloc`?

**Comment by kripken** [comment, marked as answer]
Technically you can also free the pointer from `stackAlloc` (using `stackRestore`). However, the ABI used by clang and Emscripten for the wasm stack ensures that the stack is cleaned up when the parent unwinds anyhow.

Concretely, imagine we enter a function. It stores the current stack pointer as `temp`. We can then do various small allocations in that function, and do not need to free anything there, because when the function exits it will do `stackRestore(temp)`, which in effect frees all allocations in the middle.

**Reply by cemalgnlts** [reply]
I am trying to port a graphics library. When I examined the code of someone who did this, he was using `cwrap` to add text to the screen. If I understand correctly, this function uses stringToUTF8OnStack and the part that confuses me is that this person uses `requestAnimationFrame`. So ~60 times per second it would run `cwrap`, add the string to the stack and send it to the C function.

Now I understand that this will not be a problem since space is allocated in the stack, but wouldn't it be more useful in the long run to keep the string in memory by using malloc instead?

Because I want to change my usage according to this answer.

**Reply by kripken** [reply]
If you can do a single malloc up front, and keep reusing that between calls, then yes, that sounds more efficient.

**Reply by cemalgnlts** [reply]
Thanks a lot. I understand better now.

**Reply by sbc100** [reply]
The cost of stack allocation is effectively zero for native code and very small for JS coded.  Doing ~60 stack allocations per frame should be fine.

So I would way either pre-allocating using `malloc` or using stack allocation should both be fine.  The thing you might want to avoid is calling malloc and free each frame.

**Reply by cemalgnlts** [reply]
@sbc100 Great, thanks for the explanation.
